File: 777.在-lr-字符串中交换相邻字符.js
/*
 * @lc app=leetcode.cn id=777 lang=javascript
 *
 * [777] 在 LR 字符串中交换相邻字符
 */

// @lc code=start
/**
 * @param {string} start
 * @param {string} result
 * @return {boolean}
 */
var canTransform = function(start, result) {
    
    const n = start.length;
    let i = 0, j = 0;
    
    while (i < n || j < n) {
        // 1. 跳过 start 中的所有 'X'
        while (i < n && start[i] === 'X') {
            i++;
        }
        // 2. 跳过 result 中的所有 'X'
        while (j < n && result[j] === 'X') {
            j++;
        }
        
        // 3. 如果其中一个指针到头了，另一个也必须到头
        if (i === n || j === n) {
            return i === n && j === n;
        }
        
        // 4. 相对顺序检查：当前非 'X' 字符必须相同
        if (start[i] !== result[j]) {
            return false;
        }
        
        // 5. 移动方向检查
        // 'L' 只能向左移动，所以 start 中的索引 i 必须大于或等于 result 中的索引 j
        if (start[i] === 'L' && i < j) {
            return false;
        }
        // 'R' 只能向右移动，所以 start 中的索引 i 必须小于或等于 result 中的索引 j
        if (start[i] === 'R' && i > j) {
            return false;
        }
        
        // 6. 匹配成功，两个指针同时后移
        i++;
        j++;
    }
    
    return true;
};
// @lc code=end

File: 777.在-lr-字符串中交换相邻字符.java
/*
 * @lc app=leetcode.cn id=777 lang=java
 *
 * [777] 在 LR 字符串中交换相邻字符
 */

// @lc code=start
class Solution {
    public boolean canTransform(String start, String result) {
        
        int n = start.length();
        int i = 0, j = 0;
        
        while (i < n || j < n) {
            // 1. 忽略 start 中的 'X'
            while (i < n && start.charAt(i) == 'X') {
                i++;
            }
            // 2. 忽略 result 中的 'X'
            while (j < n && result.charAt(j) == 'X') {
                j++;
            }
            
            // 如果其中一个遍历完了，另一个也必须遍历完（因为 L 和 R 的数量必须相等）
            if (i == n || j == j) {
                return i == n && j == n;
            }
            
            // 3. 此时 i 和 j 指向的都不是 'X'，检查字符是否相同
            if (start.charAt(i) != result.charAt(j)) {
                return false;
            }
            
            // 4. 检查移动方向是否合法
            // 如果是 'L'，在 start 中的位置 i 不能小于在 result 中的位置 j
            if (start.charAt(i) == 'L' && i < j) {
                return false;
            }
            // 如果是 'R'，在 start 中的位置 i 不能大于在 result 中的位置 j
            if (start.charAt(i) == 'R' && i > j) {
                return false;
            }
            
            // 5. 匹配成功，继续看下一个字符
            i++;
            j++;
        }
        
        return true;
    }
}
// @lc code=end

